offchain/advance-runner/src/dapp_contract.rs

```diff
@@ -0,0 +1,46 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0 (see LICENSE)
+
+use contracts::cartesi_dapp::CartesiDApp;
+use ethers::{
+    prelude::ContractError,
+    providers::{Http, HttpRateLimitRetryPolicy, Provider, RetryClient},
+};
+use rollups_events::{Address, Hash};
+use snafu::{ResultExt, Snafu};
+use std::sync::Arc;
+use url::Url;
+
+const MAX_RETRIES: u32 = 10;
+const INITIAL_BACKOFF: u64 = 1000;
+
+#[derive(Debug, Snafu)]
+pub enum DappContractError {
+    #[snafu(display("failed to obtain hash from dapp contract"))]
+    ContractError {
+        source: ContractError<Provider<RetryClient<Http>>>,
+    },
+}
+
+pub async fn get_template_hash(
+    dapp_address: Address,
+    provider_http_endpoint: Url,
```

Comment:
Couldn't these be references?

Reply:
`dapp_address ` can be a reference no problem. But `Url` will be consumed in the function. Is it better/more idiomatic to clone it before passing it as a parameter or cloning the reference inside?

Reply:
If it is consumed, it is better to receive a value.